In [ ]:
! apt install python3.10-venv

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  python3-pip-whl python3-setuptools-whl
The following NEW packages will be installed:
  python3-pip-whl python3-setuptools-whl python3.10-venv
0 upgraded, 3 newly installed, 0 to remove and 35 not upgraded.
Need to get 2,473 kB of archives.
After this operation, 2,884 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 python3-pip-whl all 22.0.2+dfsg-1ubuntu0.4 [1,680 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 python3-setuptools-whl all 59.6.0-1.2ubuntu0.22.04.1 [788 kB]
Get:3 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 python3.10-venv amd64 3.10.12-1~22.04.3 [5,716 B]
Fetched 2,473 kB in 1s (1,688 kB/s)
Selecting previously unselected package python3-pip-whl.
(Reading database ... 121749 files and directories currently installed.)
Pr

In [ ]:
! python -m venv env

In [ ]:
! source env/bin/activate

In [ ]:
! pip install deeppavlov

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 489.4/489.4 kB 8.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.8/55.8 kB 6.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.1/17.1 MB 47.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 83.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 222.4/222.4 kB 33.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.5/26.5 MB 42.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.5/78.5 kB 10.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.4/57.4 kB 9.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.4/34.4 MB 45.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.3/64.3 kB 10.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 10.2 MB/s eta 0:00:00
  Attempting uninstall: tqdm
    Found existing installation: tqdm 4.66.2
    Uninstalling tqdm-4.66.2:
      

In [ ]:
!python -m deeppavlov install morpho_ru_syntagrus_bert

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 887.5/887.5 MB 1.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 849.3/849.3 kB 68.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 557.1/557.1 MB 3.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 317.1/317.1 MB 3.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.0/21.0 MB 43.9 MB/s eta 0:00:00
  Attempting uninstall: torch
    Found existing installation: torch 2.1.0+cu121
    Uninstalling torch-2.1.0+cu121:
      Successfully uninstalled torch-2.1.0+cu121
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torchaudio 2.1.0+cu121 requires torch==2.1.0, but you have torch 1.13.1 which is incompatible.
torchdata 0.7.0 requires torch==2.1.0, but you have torch 1.13.1 which is incompatible.
torchtext 0.16.0 requires torch==2.1.0, but you have torch 1.13.1 

In [ ]:
!python -m spacy download ru_core_news_sm

2024-03-11 15:02:32.747515: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-03-11 15:02:32.747577: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-03-11 15:02:32.748860: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-03-11 15:02:32.756763: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-03-11 15:02:34.988492: W tensorflow/comp

In [ ]:
! pip install Natasha

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.4/34.4 MB 29.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.5/55.5 kB 6.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.7/46.7 kB 6.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.2/8.2 MB 106.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for docopt: filename=docopt-0.6.2-py2.py3-none-any.whl size=13706 sha256=c0b1603f0a1cdc4c1954c6f4a2f63a612b715674ccd89e0aa1ec8bae1a726818
  Stored in directory: /root/.cache/pip/wheels/fc/ab/d4/5da2067ac95b36618c629a5f93f809425700506f72c9732fac
  Created wheel for intervaltree: filename=intervaltree-3.1.0-py2.py3-none-any.whl size=26095 sha256=eb62fcee40baae82c335a890c5918521711124482b6edc4e9284fe26d3069af6
  Stored in directory: /root/.cache/pip/wheels/fa/80/8c/43488a924a046b733b64de3fac99252674c892a4c3801c0a61
Successfully built docopt intervaltree


In [ ]:
! pip install stanza

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 970.4/970.4 kB 10.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 421.5/421.5 kB 15.5 MB/s eta 0:00:00


In [ ]:
! pip install -U spacy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.6/6.6 MB 22.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 922.3/922.3 kB 39.3 MB/s eta 0:00:00
  Attempting uninstall: thinc
    Found existing installation: thinc 8.1.12
    Uninstalling thinc-8.1.12:
      Successfully uninstalled thinc-8.1.12
  Attempting uninstall: spacy
    Found existing installation: spacy 3.5.4
    Uninstalling spacy-3.5.4:
      Successfully uninstalled spacy-3.5.4
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
en-core-web-sm 3.5.0 requires spacy<3.6.0,>=3.5.0, but you have spacy 3.7.4 which is incompatible.
ru-core-news-sm 3.5.0 requires spacy<3.6.0,>=3.5.0, but you have spacy 3.7.4 which is incompatible.


# Сравнение теггеров

In [ ]:
import json
import pandas as pd
from sklearn.metrics import accuracy_score
from nltk import word_tokenize
import nltk
nltk.download('punkt')
from string import punctuation
punctuation += '—'+'–'

import stanza
stanza.download('ru')

stnz = stanza.Pipeline('ru', use_gpu = True)
stnz_cpu = stanza.Pipeline('ru', use_gpu = False)

import spacy
spacy.prefer_gpu()
spc = spacy.load("ru_core_news_sm")

from natasha import (
    Segmenter,
    MorphVocab,
    NewsEmbedding,
    NewsMorphTagger,
    Doc
)

segmenter = Segmenter()
morph_vocab = MorphVocab()
emb = NewsEmbedding()
morph_tagger = NewsMorphTagger(emb)

from pymystem3 import Mystem

m = Mystem()

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Downloading default packages for language: ru (Russian) ...


INFO:stanza:Downloaded file to /root/stanza_resources/ru/default.zip
INFO:stanza:Finished downloading models and saved to /root/stanza_resources
INFO:stanza:Checking for updates to resources.json in case models have been updated.  Note: this behavior can be turned off with download_method=None or download_method=DownloadMethod.REUSE_RESOURCES


INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Loading these models for language: ru (Russian):
| Processor | Package            |
----------------------------------
| tokenize  | syntagrus          |
| pos       | syntagrus_charlm   |
| lemma     | syntagrus_nocharlm |
| depparse  | syntagrus_charlm   |
| ner       | wikiner            |

INFO:stanza:Using device: cuda
INFO:stanza:Loading: tokenize
INFO:stanza:Loading: pos
INFO:stanza:Loading: lemma
INFO:stanza:Loading: depparse
INFO:stanza:Loading: ner
INFO:stanza:Done loading processors!
INFO:stanza:Checking for updates to resources.json in case models have been updated.  Note: this behavior can be turned off with download_method=None or download_method=DownloadMethod.REUSE_RESOURCES


INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Loading these models for language: ru (Russian):
| Processor | Package            |
----------------------------------
| tokenize  | syntagrus          |
| pos       | syntagrus_charlm   |
| lemma     | syntagrus_nocharlm |
| depparse  | syntagrus_charlm   |
| ner       | wikiner            |

INFO:stanza:Using device: cpu
INFO:stanza:Loading: tokenize
INFO:stanza:Loading: pos
INFO:stanza:Loading: lemma
INFO:stanza:Loading: depparse
INFO:stanza:Loading: ner
INFO:stanza:Done loading processors!
/usr/local/lib/python3.10/dist-packages/spacy/util.py:910: UserWarning: [W095] Model 'ru_core_news_sm' (3.5.0) was trained with spaCy v3.5.0 and may not be 100% compatible with the current version (3.7.4). If you see errors or degraded performance, download a newer compatible model or retrain your custom model with the current spaCy version. For more details and available updates, run: python -m spacy validate
  warn

In [ ]:
from deeppavlov import build_model

dp = build_model("morpho_ru_syntagrus_bert", download=True, install=True)

2024-03-11 15:57:02.455 INFO in 'deeppavlov.core.data.utils'['utils'] at line 95: Downloading from http://files.deeppavlov.ai/deeppavlov_data/morpho_tagger/UD2.3/ru_syntagrus.tar.gz to /root/.deeppavlov/downloads/UD2.3_source/ru_syntagrus.tar.gz
INFO:deeppavlov.core.data.utils:Downloading from http://files.deeppavlov.ai/deeppavlov_data/morpho_tagger/UD2.3/ru_syntagrus.tar.gz to /root/.deeppavlov/downloads/UD2.3_source/ru_syntagrus.tar.gz
100%|██████████| 16.7M/16.7M [00:01<00:00, 9.27MB/s]
2024-03-11 15:57:05.284 INFO in 'deeppavlov.core.data.utils'['utils'] at line 276: Extracting /root/.deeppavlov/downloads/UD2.3_source/ru_syntagrus.tar.gz archive into /root/.deeppavlov/downloads/UD2.3_source/ru_syntagrus
INFO:deeppavlov.core.data.utils:Extracting /root/.deeppavlov/downloads/UD2.3_source/ru_syntagrus.tar.gz archive into /root/.deeppavlov/downloads/UD2.3_source/ru_syntagrus
2024-03-11 15:57:06.868 INFO in 'deeppavlov.core.data.utils'['utils'] at line 95: Downloading from http://files.

tokenizer_config.json:   0%|          | 0.00/24.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/642 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/1.65M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/714M [00:00<?, ?B/s]

Some weights of the model checkpoint at DeepPavlov/rubert-base-cased were not used when initializing BertForTokenClassification: ['cls.predictions.decoder.weight', 'cls.predictions.transform.dense.bias', 'cls.seq_relationship.bias', 'cls.predictions.transform.dense.weight', 'cls.seq_relationship.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.bias', 'cls.predictions.decoder.bias', 'cls.predictions.transform.LayerNorm.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForTokenClassification were not initializ

### Проблемы автоматической разметки:
Рассмотрим случаи, которые вызвают трудности при автоматической разработке. В корпусе будут представлены по 1-2 предложения на каждый пример.
- омонимия разного рода:
    - существительное vs глагол *печь, ели*
    - причастия vs прилагательные *вареный, рассенный*
    - наречия vs предлоги: *вблизи*
- слова с дефисами.
    - Не хотим делить:
         - предлоги: *из-за*
         - наречия: *по-настоящему*
         - названия городов, в том числе другие части речи, образованные от их названий: *Нью-Йорк*, *Йошкар-Олинский*
     - Хотим делить:
         - частицы: *-таки*
         - маршрут: *Белорецк-Магнитогорск*
         - диапазон:*5-6*
- имена собственные, как экзотические, так и фамилии на *-ов, -их*: *Тарасевича (Тарасевич), Петровский, Лидваль, Шемякин, Резвых, от Гомера до Андре Жида*
- сокращения *тыс., ред.*

* Категория "*Редкие слова*" по [Ляшевская et.al 2010]:
    * слова с неизвестным словарю корнем, но образованные с помощью продуктивных аффиксов. В моем корпусе: *симпим (симпить), заспавнилась, коупила (коупить), инфодамплю, галюники, панчи, ремув, на самике (самик), фички*
    * совсем непонятные слова. В моем корпусе: *на бэбича (бэбич), упртст, делюлю (прилагательное)*
    * окказиональные формы: *стригя, пья, победу, льзя*
    * аббривеатуры: *МГУ, ВМО, РФФИ, РАН*

* случайно взятые слова из тех, которые не сомгли распарсить в ГИКРЯ:
  * *Егошная, егошнее* - просторечия
  * *Егэшный, ВААЩе* - неологизмы и слэнг,
  * *пяЦот, пэсню* - подражание устной речи
  * *пюдям, пюока* - опечатки
  

От себя:
* *всеееееее* - удвоенные буквы
* другое: *покашт, ниче, ченить*

## Словарь универсальных тегов
В рассмотрении будут **Natasha, Stanza, Mystem, Spacy, DeepPavlov**

Natasha, Stanza и Spacy используют Upos, у Mystem свои теги. Сравним, как это будет работать для разных частей речи и подберем один враиант

| часть речи | тег Mystem | Upos | Наш итог |
|:----------:|:----------:|:---------:|:--------:|
| прилагательное | A | ADJ | **ADJ** |
| наречие | ADV | ADV | **ADV** |
| местоименное наречие | ADVPRO | ADV | **ADV** |
| числительное-прилагательное | ANUM | ADJ | **ADJ** |
| местоимение-прилагательное | APRO | DET | **DET** |
| часть композита - сложного слова | COM | ? | **NOUN** |
| сочинительный союз | CONJ | CCONJ | **CONJ** |
| подчинительный союз | CONJ | SCONJ | **CONJ** |
| междометие | INTJ | INTJ | **INTJ** |
| числительное | NUM | NUM | **NUM** |
| частица | PART | PART | **PART** |
| предлог| PR | ADP | **ADP** |
| существительное| S | NOUN | **NOUN** |
| местоимение-существительное| SPRO | PRON | **PRON** |
| глагол | V | VERB | **VERB** |
| вспомогательный глагол | V | AUX | **VERB**
| имя собственное | S | PROPN | **NOUN** |
| пунктуация | - | PUNCT | **NA** |
| символ | - | SYM | **NA** |



Итог: в основном теги в Stanza и Natasha останутся нетронутыми. Имена собственные вспомогательные глаголы и союзы будут более обощены, а символы и пункутуация вовсе не будут учитываться.


In [ ]:
tokens_frame = pd.DataFrame(columns = ['Токен', 'Тег_1', 'Тег_2']) # второй тег для неоднозначных по моему мнению случаев

Использую токенизацию с помощью nltk для удобства и превращаю корпус в таблицу. Далее надо разметить вручную все токены, проверить правильность токенизации

In [ ]:
from google.colab import drive

In [ ]:
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
with open('/content/drive/My Drive/писько/Сравнение парсеров/corpora.txt', 'r', encoding ='utf-8') as file:
    raw_text = file.read()
    tokens_text = word_tokenize(raw_text)
    tokens_frame['Токен'] = tokens_text

In [ ]:
#tokens_frame.to_csv('/content/drive/My Drive/pisko/empty_tokens.csv')

In [ ]:
# подгружаю размеченные вручную токены
gold_taggs = pd.read_csv('/content/drive/My Drive/писько/Сравнение парсеров/tagged_lemmas.csv')

In [ ]:
gold_taggs

Токен Тег_1 Тег_2    лемма лемма 2
0    Входим  VERB   NaN    войти     NaN
1          в   ADP   NaN        в     NaN
2        дом  NOUN   NaN      дом     NaN
3          ,   NaN   NaN        ,     NaN
4     открыв  VERB   NaN  открыть     NaN
..       ...   ...   ...      ...     ...
666    пюока  PART   NaN     пока     NaN
667        ,   NaN   NaN        ,     NaN
668       до   ADP   NaN       до     NaN
669  скорова   ADJ   NaN   скорый     NaN
670        !   NaN   NaN        !     NaN

[671 rows x 5 columns]

In [ ]:
print('Размер корпуса: ', len(gold_taggs))

Размер корпуса:  671


## Задаем функции

In [ ]:
upos_tags = {'CCONJ': 'CONJ',
             'SCONJ': 'CONJ',
             'AUX': 'VERB',
             'PROPN': 'NOUN',
             'PUNCT': '',
             'SYM': ''
}

In [ ]:
mystem_tags = {'A': 'ADJ',
               'ADVPRO': 'ADV',
               'ANUM': 'ADJ',
               'APRO': 'DET',
               'COM': 'NOUN',
               'PR': 'ADP',
               'S': 'NOUN',
               'SPRO': 'PRON',
               'V': 'VERB',
              }

In [ ]:
def compare_right_tokens(token, tag, lemma, gold, results_tag, results_lemma, gold_tags, gold_lemmas, index):
    tag_gold_1 = gold['Тег_1'][index]
    tag_gold_2 = gold['Тег_2'][index]
    lemma_gold_1 = str(gold['лемма'][index]).lower()
    lemma_gold_2 = str(gold['лемма 2'][index]).lower()

    if tag == tag_gold_2:
         gold_tags.append(tag_gold_2)
    else: gold_tags.append(tag_gold_1)

    if lemma.lower() == lemma_gold_2.lower():
         gold_lemmas.append(lemma_gold_2)
    else: gold_lemmas.append(lemma_gold_1)
    results_tag.append(tag)
    results_lemma.append(lemma.lower())


In [ ]:
def compare(gold, unified_parser):
    results_tag = []  # сюда будем писать результаты
    results_lemma = []
    gold_tags = []     # сюда будем писать мои теги
    gold_lemmas = []
    index = 0
    for token, tag, lemma in zip(unified_parser['Токен'], unified_parser['Тег'], unified_parser['Лемма']):
        token_gold = gold['Токен'][index]
        tag_gold_1 = gold['Тег_1'][index]
        lemma_gold_1 = str(gold['лемма'][index]).lower()

        # если все нормально с токинизацией
        if token == token_gold:
            res = compare_right_tokens(token, tag, lemma, gold, results_tag, results_lemma, gold_tags, gold_lemmas, index)
            index += 1

        # если неправильная токенизация
        else:
            if token in str(token_gold): # парсер побил на токены лишний раз
                print('Парсер перебил ' + str(token_gold) + '. Выдал: ' + str(token))

                if token_gold.endswith(token):
                    results_tag.append('')
                    results_lemma.append('')
                    gold_tags.append(tag_gold_1)
                    gold_lemmas.append(lemma_gold_1)
                    index += 1
                else: continue # без прибавления счетчика, пока не дойдем до конца токена

            elif str(token_gold) in token: # парсер недобил
                print('Парсер недобил ' + str(token_gold) + '. Выдал: ' + str(token))
                while not token.endswith(token_gold):
                    token_gold = gold['Токен'][index]
                    lemma_gold = gold['лемма'][index]
                    index += 1
                results_tag.append('')
                results_lemma.append('')
                gold_lemmas.append(str(gold['лемма'][index]).lower())
                gold_tags.append(gold['Тег_1'][index])

            else:
                print('Не совпадают:', token_gold, token)

    print('Точность по тегам:')
    print(accuracy_score(results_tag, gold_tags))
    print('Точность по леммам:')
    print(accuracy_score(results_lemma, gold_lemmas))

In [ ]:
# создаем таблицу с разметкой теггера
def create_tagged_frame(tag_dict):
    new_frame = pd.DataFrame(columns = ['Токен', 'Тег', 'Лемма'], index = range(0, len(tag_dict)))
    tokens = [token[0] for token in tag_dict]
    tags = [tag[1] for tag in tag_dict]
    lemmas = [lemma[2] for lemma in tag_dict]

    new_frame['Токен'] = tokens
    new_frame['Тег'] = tags
    new_frame['Лемма'] = lemmas

    return new_frame

In [ ]:
def unification(dct):
    unified = []
    for token_n_tag in dct:
        token = token_n_tag[0]
        tag = token_n_tag[1]
        lemma = token_n_tag[2]
        if tag in mystem_tags:
            unified.append((token, mystem_tags[tag], lemma))
        elif tag in upos_tags:
            unified.append((token, upos_tags[tag], lemma))
        else: unified.append((token, tag, lemma))
    return create_tagged_frame(unified)


# Stanza

In [ ]:
%%time
stanza_dict = [] # на самом деле не словарь, а лист коретежей, чтобы значения могли повторяться
analys = stnz_cpu(raw_text).to_dict()
for i in range(0, len(analys)):
    for sent in analys[i]:
        stanza_dict.append((sent['text'], sent['upos']))

KeyboardInterrupt: 

In [ ]:
%%time
stanza_dict = [] # на самом деле не словарь, а лист коретежей, чтобы значения могли повторяться
analys = а.to_dict()
for i in range(0, len(analys)):
    for sent in analys[i]:
        if sent['lemma']:
          lemma = sent['lemma']
        else: lemma = None
        stanza_dict.append((sent['text'], sent['upos'], lemma))

CPU times: user 1.14 s, sys: 87.3 ms, total: 1.22 s
Wall time: 1.22 s


In [ ]:
stanza_dict[:10]

[('\ufeffВходим', 'VERB', 'свходить'),
 ('в', 'ADP', 'в'),
 ('дом', 'NOUN', 'дом'),
 (',', 'PUNCT', ','),
 ('открыв', 'VERB', 'открыть'),
 ('дверь', 'NOUN', 'дверь'),
 (',', 'PUNCT', ','),
 ('попадаем', 'VERB', 'попадать'),
 ('в', 'ADP', 'в'),
 ('кухню', 'NOUN', 'кухня')]

In [ ]:
stanza_frame = unification(stanza_dict)

In [ ]:
stanza_frame

Токен   Тег     Лемма
0    Входим  VERB  свходить
1          в   ADP         в
2        дом  NOUN       дом
3          ,               ,
4     открыв  VERB   открыть
..       ...   ...       ...
659    пюока  VERB   пюокать
660        ,               ,
661       до   ADP        до
662  скорова  NOUN   скорова
663        !               !

[664 rows x 3 columns]

In [ ]:
results_lemma, gold_lemmas = compare(gold_taggs, stanza_frame)

Парсер недобил Немцы. Выдал: Немцы-таки
Парсер недобил Белорецк. Выдал: Белорецк-Магнитогорск
Парсер недобил тыс. Выдал: тыс.
Парсер недобил Прим. Выдал: Прим.
Парсер недобил ред. Выдал: ред.
Точность по тегам:
0.7469879518072289
Точность по леммам:
0.9081325301204819


# Natasha

In [ ]:
%%time
natasha_dict = [] # на самом деле не слвоарь, а лист коретежей, чтобы значения могли повторяться
doc = Doc(raw_text)
doc.segment(segmenter)
doc.tag_morph(morph_tagger)

for token in doc.tokens:
    token.lemmatize(morph_vocab)
    natasha_dict.append((token.text, token.pos, token.lemma))

CPU times: user 111 ms, sys: 53.2 ms, total: 164 ms
Wall time: 92.5 ms


In [ ]:
natasha_dict[:10]

[('\ufeffВходим', 'PUNCT', '\ufeffвходим'),
 ('в', 'ADP', 'в'),
 ('дом', 'NOUN', 'дом'),
 (',', 'PUNCT', ','),
 ('открыв', 'VERB', 'открыть'),
 ('дверь', 'NOUN', 'дверь'),
 (',', 'PUNCT', ','),
 ('попадаем', 'VERB', 'попадать'),
 ('в', 'ADP', 'в'),
 ('кухню', 'NOUN', 'кухня')]

In [ ]:
natasha_frame = unification(natasha_dict)

In [ ]:
natasha_frame.head()

Токен   Тег    Лемма
0  Входим        входим
1        в   ADP        в
2      дом  NOUN      дом
3        ,              ,
4   открыв  VERB  открыть

In [ ]:
compare(gold_taggs, natasha_frame)

Парсер недобил 5. Выдал: 5-6
Парсер недобил Немцы. Выдал: Немцы-таки
Парсер перебил А.. Выдал: А
Парсер перебил А.. Выдал: .
Парсер перебил Е.. Выдал: Е
Парсер перебил Е.. Выдал: .
Парсер недобил Белорецк. Выдал: Белорецк-Магнитогорск
Парсер недобил 2001. Выдал: 2001-2002
Точность по тегам:
0.7375565610859729
Точность по леммам:
0.9140271493212669


# Mystem

In [ ]:
%%time
doc = m.analyze(raw_text)

CPU times: user 5.42 ms, sys: 2.24 ms, total: 7.65 ms
Wall time: 54.9 ms


Заметим, что работает в разы дольше соперниц

In [ ]:
doc[1]['analysis'][0]['lex']

'входить'

In [ ]:
mystem_dict = [] # на самом деле не слвоарь, а лист коретежей, чтобы значения могли повторяться
for i in range(0, len(doc)):
    txt = doc[i]['text'].replace(' ', '') #приклеивает пробелы к пунктуации
    try:
        mystem_dict.append((txt,
                            doc[i]['analysis'][0]['gr'].split('=')[0].split(',')[0],
                           doc[i]['analysis'][0]['lex']))
    except Exception as e:
        #print(e)
        if txt.isalpha() or txt.isdigit(): #пропускает незнакомые слова, числа и пунктуацию, что плохо
            mystem_dict.append((txt, ' ', ' '))
        else:
            if (txt in punctuation or txt[0] in punctuation) and txt != '':
                mystem_dict.append((txt, ' ', ' '))

In [ ]:
mystem_dict[:10]

[('Входим', 'V', 'входить'),
 ('в', 'PR', 'в'),
 ('дом', 'S', 'дом'),
 (',', ' ', ' '),
 ('открыв', 'V', 'открывать'),
 ('дверь', 'S', 'дверь'),
 (',', ' ', ' '),
 ('попадаем', 'V', 'попадать'),
 ('в', 'PR', 'в'),
 ('кухню', 'S', 'кухня')]

In [ ]:
mystem_frame = unification(mystem_dict)

In [ ]:
compare(gold_taggs, mystem_frame)

Парсер перебил Входим. Выдал: Входим
Парсер перебил А.. Выдал: А
Парсер перебил А.. Выдал: .
Парсер перебил Е.. Выдал: Е
Парсер перебил Е.. Выдал: .
Парсер перебил 15,5. Выдал: 15
Парсер перебил 15,5. Выдал: ,
Парсер перебил 15,5. Выдал: 5
Парсер недобил .. Выдал: .):
Точность по тегам:
0.7339312406576981
Точность по леммам:
0.7085201793721974


# Spacy

In [ ]:
import locale
locale.getpreferredencoding = lambda: "UTF-8"

In [ ]:
!pip install -U 'spacy[cuda11]'

In [ ]:
%%time
spc_dict = [] # на самом деле не слвоарь, а лист коретежей, чтобы значения могли повторяться
doc = spc(raw_text)
for token in doc:
  if token.text != '\n':
    spc_dict.append((token.text, token.pos_, token.lemma_))

CPU times: user 254 ms, sys: 0 ns, total: 254 ms
Wall time: 255 ms


In [ ]:
spc_dict[:10]

[('\ufeffВходим', 'PROPN', '\ufeffвходим'),
 ('в', 'ADP', 'в'),
 ('дом', 'NOUN', 'дом'),
 (',', 'PUNCT', ','),
 ('открыв', 'VERB', 'открыть'),
 ('дверь', 'NOUN', 'дверь'),
 (',', 'PUNCT', ','),
 ('попадаем', 'VERB', 'попадать'),
 ('в', 'ADP', 'в'),
 ('кухню', 'NOUN', 'кухня')]

In [ ]:
spc_frame = unification(spc_dict)

In [ ]:
compare(gold_taggs, spc_frame)

Парсер перебил из-за. Выдал: из
Парсер перебил из-за. Выдал: -
Парсер перебил из-за. Выдал: за
Парсер перебил из-за. Выдал: из
Парсер перебил из-за. Выдал: -
Парсер перебил из-за. Выдал: за
Парсер перебил по-настоящему. Выдал: по
Парсер перебил по-настоящему. Выдал: -
Парсер перебил по-настоящему. Выдал: настоящему
Парсер перебил Йошкар-Олинском. Выдал: Йошкар
Парсер перебил Йошкар-Олинском. Выдал: -
Парсер перебил Йошкар-Олинском. Выдал: Олинском
Парсер перебил Нью-Йорке. Выдал: Нью
Парсер перебил Нью-Йорке. Выдал: -
Парсер перебил Нью-Йорке. Выдал: Йорке
Парсер перебил какой-то. Выдал: какой
Парсер перебил какой-то. Выдал: -
Парсер перебил какой-то. Выдал: то
Не совпадают: Заболел   
Парсер недобил тыс. Выдал: тыс.
Парсер недобил ). Выдал: ):
Парсер перебил когда-нибудь. Выдал: когда
Парсер перебил когда-нибудь. Выдал: -
Парсер перебил когда-нибудь. Выдал: нибудь
Точность по тегам:
0.7369207772795217
Точность по леммам:
0.852017937219731


# DeepPavlov

In [ ]:
! pip install conllu

In [ ]:
 from conllu import parse

In [ ]:
%%time
dp_doc = dp(raw_text.split('\n')) #не может сам бить

CPU times: user 4.43 s, sys: 4.76 ms, total: 4.44 s
Wall time: 5.63 s


In [ ]:
dp_doc[:3]

['1\t\ufeffВходим\t\ufeffВходим\tVERB\t_\tAspect=Imp|Mood=Ind|Number=Plur|Person=1|Tense=Pres|VerbForm=Fin|Voice=Act\t_\t_\t_\t_\n2\tв\tв\tADP\t_\t_\t_\t_\t_\t_\n3\tдом\tдом\tNOUN\t_\tAnimacy=Inan|Case=Acc|Gender=Masc|Number=Sing\t_\t_\t_\t_\n4\t,\t,\tPUNCT\t_\t_\t_\t_\t_\t_\n5\tоткрыв\tоткрыть\tVERB\t_\tAspect=Perf|Tense=Past|VerbForm=Conv|Voice=Act\t_\t_\t_\t_\n6\tдверь\tдверь\tNOUN\t_\tAnimacy=Inan|Case=Acc|Gender=Fem|Number=Sing\t_\t_\t_\t_\n7\t,\t,\tPUNCT\t_\t_\t_\t_\t_\t_\n8\tпопадаем\tпопадать\tVERB\t_\tAspect=Imp|Mood=Ind|Number=Plur|Person=1|Tense=Pres|VerbForm=Fin|Voice=Act\t_\t_\t_\t_\n9\tв\tв\tADP\t_\t_\t_\t_\t_\t_\n10\tкухню\tкухня\tNOUN\t_\tAnimacy=Inan|Case=Acc|Gender=Fem|Number=Sing\t_\t_\t_\t_\n11\t:\t:\tPUNCT\t_\t_\t_\t_\t_\t_\n12\tпечь\tпечь\tNOUN\t_\tAnimacy=Inan|Case=Nom|Gender=Fem|Number=Sing\t_\t_\t_\t_\n13\t,\t,\tPUNCT\t_\t_\t_\t_\t_\t_\n14\tналево\tналево\tADV\t_\tDegree=Pos\t_\t_\t_\t_\n15\tокно\tокно\tNOUN\t_\tAnimacy=Inan|Case=Nom|Gender=Neut|Number=Sing\t_\

In [ ]:
dp_dict = [] # на самом деле не слвоарь, а лист коретежей, чтобы значения могли повторяться
for line in dp_doc:
  for token in parse(line)[0]:
    dp_dict.append((token['form'], token['upos'], token['lemma']))

In [ ]:
dp_dict[:10]

[('\ufeffВходим', 'VERB', '\ufeffВходим'),
 ('в', 'ADP', 'в'),
 ('дом', 'NOUN', 'дом'),
 (',', 'PUNCT', ','),
 ('открыв', 'VERB', 'открыть'),
 ('дверь', 'NOUN', 'дверь'),
 (',', 'PUNCT', ','),
 ('попадаем', 'VERB', 'попадать'),
 ('в', 'ADP', 'в'),
 ('кухню', 'NOUN', 'кухня')]

In [ ]:
dp_frame = unification(dp_dict)

In [ ]:
compare(gold_taggs, dp_frame)

Парсер недобил 5. Выдал: 5-6
Парсер недобил Немцы. Выдал: Немцы-таки
Парсер недобил Белорецк. Выдал: Белорецк-Магнитогорск
Парсер недобил 2001. Выдал: 2001-2002
Точность по тегам:
0.7541478129713424
Точность по леммам:
0.8235294117647058


### Итог

|| Stanza | Natasha | MyStem | Spacy | DeepPavlov |
|:----------:|:----------:|:----------:|:---------:|:--------:|:--------:|
|точность тег|0.745 | 0.736 | 0.732 | 0.735 | 0.752 |
|точность лемма|0.908 | 0.914 | 0.708 | 0.85| 0.823 |
|время CPU|17.5 s|136 ms|1.13 s| 615 ms|11.6 s|
|время GPU|1.22 s|92.5 ms|54.9 ms| 255 ms|4 s|


Хотя DeepPavlov самый точный, он работает слишком долго.
Самый быстрый парсер - Natasha, по точности превосходит MyStem и Spacy